lib/automation.go

```diff
@@ -126,3 +147,183 @@ func (automationImpl) Apply(scope scope, p *ApplyParams) (*ApplyResult, error) {
+		Apply:   false,
+	}
+	// TODO(ramfox): bandaid! remove when save can handle having saving a dataset with no body/structure
+	if p.Dataset.ID == "" && p.Dataset.BodyFile() == nil {
+		saveParams.Apply = true
+	}
+	ds, err := datasetImpl{}.Save(scope, saveParams)
+	if err != nil && !errors.Is(err, dsfs.ErrNoChanges) {
+		log.Errorw("deploy save dataset", "error", err)
+		deployPayload.Error = err.Error()
+		go func(dp *event.DeployEvent) {
+			if err := scope.Bus().PublishID(scope.Context(), event.ETDeployEnd, ref, deployPayload); err != nil {
+				log.Debug(err)
+			}
+		}(&deployPayload)
+		return
+	}
+
+	deployPayload.DatasetID = ds.ID
+	go func(dp *event.DeployEvent) {
+		if err := scope.Bus().PublishID(scope.Context(), event.ETDeploySaveDatasetEnd, ref, deployPayload); err != nil {
+			log.Debug(err)
+		}
+	}(&deployPayload)
+
+	wf := p.Workflow.Copy()
+	wf.DatasetID = ds.ID
+	wf.OwnerID = scope.ActiveProfile().ID
+
+	go func(dp *event.DeployEvent) {
+		if err := scope.Bus().PublishID(scope.Context(), event.ETDeploySaveWorkflowStart, ref, deployPayload); err != nil {
+			log.Debug(err)
+		}
+	}(&deployPayload)
+
+	wf, err = scope.Automation().SaveWorkflow(wf)
+	if err != nil {
+		log.Errorw("deploy save workflow", "error", err)
+		deployPayload.Error = err.Error()
+		go func(dp *event.DeployEvent) {
+			if err := scope.Bus().PublishID(scope.Context(), event.ETDeployEnd, ref, deployPayload); err != nil {
+				log.Debug(err)
+			}
+		}(&deployPayload)
+		return
+	}
+
+	deployPayload.WorkflowID = wf.ID.String()
+	go func(dp *event.DeployEvent) {
+		if err := scope.Bus().PublishID(scope.Context(), event.ETDeploySaveWorkflowEnd, ref, deployPayload); err != nil {
+			log.Debug(err)
+		}
+	}(&deployPayload)
+
+	if p.Run {
+		runID := run.NewID()
+
+		deployPayload.RunID = runID
+		go func(dp *event.DeployEvent) {
+			if err := scope.Bus().PublishID(scope.Context(), event.ETDeployRun, ref, deployPayload); err != nil {
+				log.Debug(err)
+			}
+		}(&deployPayload)
+
+		err := scope.Automation().RunWorkflow(scope.Context(), wf.ID, runID)
+		if err != nil {
+			log.Errorw("deploy run workflow", "error", err)
+			deployPayload.Error = err.Error()
+			go func(dp *event.DeployEvent) {
+				if err := scope.Bus().PublishID(scope.Context(), event.ETDeployEnd, ref, deployPayload); err != nil {
+					log.Debug(err)
+				}
+			}(&deployPayload)
+			return
+		}
+
+	}
+	log.Debug("deploy ended")
+	go func(dp *event.DeployEvent) {
+		if err := scope.Bus().PublishID(scope.Context(), event.ETDeployEnd, ref, deployPayload); err != nil {
+			log.Debug(err)
+		}
+	}(&deployPayload)
+	rollback = false
+}
+
+func (inst *Instance) run(ctx context.Context, streams ioes.IOStreams, w *workflow.Workflow, runID string) error {
+	ctxWithProfile := profile.AddToContext(ctx, w.OwnerID.String())
+	scope, err := newScope(ctxWithProfile, inst, "local")
+	if err != nil {
+		return err
+	}
+	ref, err := scope.ResolveReference(ctx, &dsref.Ref{InitID: w.DatasetID})
```

Comment:
Bug here: the `ref` variable is actually the "source" that the reference resolved from, which will always be `""` b/c the scope uses a local resolver.  Instead use the `Human()` method on the reference passed into `ResolveReference`:

```go
ref := &dsref.Ref{InitID: w.DatasetID}
	if _, err = scope.ResolveReference(ctx, ref); err != nil {
		return fmt.Errorf("resolving InitID %q: %w", w.DatasetID, err)
	}
	p := &SaveParams{
		Ref: ref.Human(),
		Dataset: &dataset.Dataset{
			ID: w.DatasetID,
			Commit: &dataset.Commit{
				RunID: runID,
			},
		},
```

be sure to rebase from master to make sure you're getting the right error.

Reply:
yea saw this and force pushed when I looked with fresh eyes